File: overload_demo.rb
require 'java'

java_import 'OverloadDemo'

demo = OverloadDemo.new

puts demo.what_type_is 42
puts demo.java_send :whatTypeIs, [Java::int], 42
puts demo.what_type_is "Fun!"
puts demo.what_type_is Hash.new

File: OverloadDemo.java
import java.util.List;

public class OverloadDemo {
  public String whatTypeIs(long value) {
    return "long";
  }

  public String whatTypeIs(int value) {
    return "int";
  }

  public String whatTypeIs(String value) {
    return "string";
  }

  public String whatTypeIs(Object value) {
    return "object";
  }

}